compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -840,6 +840,117 @@ impl<'hir> LoweringContext<'_, 'hir> {
         })
     }
 
+    /// Destructure the LHS of complex assignments.
+    /// For instance, lower `(a, b) = t` to `{ let (lhs1, lhs2) = t; a = lhs1; b = lhs2; }`.
+    fn lower_expr_assign(
+        &mut self,
+        lhs: &Expr,
+        rhs: &Expr,
+        eq_sign_span: Span,
+        whole_span: Span,
+    ) -> hir::ExprKind<'hir> {
+        // Return early in case of an ordinary assignment.
+        let is_ordinary = match &lhs.kind {
+            ExprKind::Tup(..) => false,
+            // `(..)`.
+            ExprKind::Paren(e) => {
```

Comment:
I think it would be surprising if this didn't work. Can you think of a situation this might create undesirable behaviour? The intent of the RFC was to match the syntax for patterns as closely as possible.

Reply:
This case has no practical value, but has to be specifically implemented.
Also, in other cases parentheses disable destructuring, but in this case they enable it.
Also, this case can be implemented backward-compatibly at any later point in time.
Seems like enough reasons to not implement it now.

Reply:
I think surrounding in parentheses should not disable destructuring (but should trigger the unnecessary parenthesisation lint), which matches how patterns work. (That is, I think this implementation should be tweaked to ignore parentheses.)

> Also, this case can be implemented backward-compatibly at any later point in time.

That's not an argument against implementing this component, because it's true generally for all of destructuring assignment. The question is whether we want it or not, and considering we do allow `(..)` in patterns, there's a strong argument for supporting it.

Reply:
I agree with @varkor. The "shielding behavior" of parentheses is a bug and I'll fix it. 

> (Yes, there's a symmetry with `(..)` in patterns, but I'm not sure supporting that in patterns was a good idea in the first place.)

Well, for better or for worse, patterns work that way now. I think it would add more inconsistency to disallow `(..)` in destructuring assignments now. The RFC doesn't make an exception for it, so I don't think we should just depart from what the RFC says.

> This case has no practical value, but has to be specifically implemented.

Implementation complexity shouldn't be a reason for inconsistent behavior. I don't see a super convincing use case for `(..)` in patterns either. A somewhat convincing use case would be usage in a macro `m!(x1, x2, x3)` generating `(x1, x2, x3, ..) = foo()` , which would otherwise have to special case the empty sequence `(..)`. A lot of constructs in Rust don't have a use case (just have a look at https://github.com/rust-lang/rust/blob/master/src/test/ui/weird-exprs.rs), but they are allowed *for consistency*.

Reply:
Many examples of useless code in `weird-exprs.rs` are accepted because we didn't add extra logic to reject them. Here we add extra logic to enable useless code.
Leaving this at your discretion.

Reply:
Given that the implementation is only 4 lines longer with the handling of `(..)` included, improves consistency in @varkor's and my eyes, and there was no exception for it in the RFC, I'd like to keep it.